### PR TITLE
feat: support VMware workstation provider

### PR DESCRIPTION
**Pull request by Gexinyu2333**
When a provider adds resources in the form of VMware on the providers interface. As shown below:
b596834870087c86e5de667c2f823e6

The Client ID is a virtual machine ID of VMware Workstation Pro belonging to the provider; Client secret is the credential for VMware Workstation Pro REST service, in the form of {username}:{password}
If the VMware Workstation Pro REST service of the provider host is enabled, all virtual machines information in the provider's VMware Workstation Pro will be displayed on the machines page. As shown in the following figure:
e53f78b1ea8c4d359878b1752c2b653

### CODE REVIEW

File: object/machine.go [hunk 1]
```diff
@@ -95,8 +95,7 @@ func GetMachines(owner string) ([]*Machine, error) {
 		if err2 != nil {
 			return machines, err2
 		}
-
-		clientMachines, err2 := client.GetMachines()
+		clientMachines, err2 := (*client).GetMachines()
```
Comment (hsluoyz) [inline object/machine.go:98]: Rebase your code onto the master HEAD

File: object/machine.go [hunk 2]
```diff
@@ -135,14 +134,13 @@ func getMachine(owner string, name string) (*Machine, error) {
 			return nil, err2
 		}
 
-		clientMachine, err2 := client.GetMachine(name)
+		clientMachine, err2 := (*client).GetMachine(name)
```
Comment (hsluoyz) [inline object/machine.go:137]: Rebase your code onto the master HEAD

File: service/aliyun.go [hunk 3]
```diff
@@ -50,17 +50,17 @@ type MachineAliyunClient struct {
 	Client *ecs.Client
 }
 
-func NewMachineAliyunClient(accessKeyId string, accessKeySecret string, region string) (*MachineAliyunClient, error) {
+func NewMachineAliyunClient(accessKeyId string, accessKeySecret string, region string) (MachineAliyunClient, error) {
```
Comment (hsluoyz) [inline service/aliyun.go]: Rebase your code onto the master HEAD

File: service/util.go [hunk 4]
```diff
@@ -30,3 +34,15 @@ func getLocalTimestamp(input string) string {
 	res := utcTime.Local().Format(time.RFC3339)
 	return res
 }
+
+func ExtractFileName(path string) string {
```
Comment (hsluoyz) [inline service/util.go]: Go std lib should have this func

File: service/vmware.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
+// you may not use this file except in compliance with the License.
```
Comment (hsluoyz) [inline service/vmware.go:4]: Not correct Apache header

File: service/vmware.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	var vmList []VirtualMachinePath
+	err = json.Unmarshal(body, &vmList)
+	if err != nil {
+		return nil, err
+	}
+	if len(vmList) == 0 {
```
Comment (hsluoyz) [inline service/vmware.go]: Remove this if

File: service/vmware.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+		return nil, nil
+	}
+	for _, vm := range vmList {
+		client.accessKeyId = vm.ID
+		name := ExtractFileName(vm.Path)
+		machine, err := client.GetMachine(name)
```
Comment (hsluoyz) [inline service/vmware.go]: Use Go routine to speed up

File: service/vmware.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+		accessKeySecret,
+		region,
+	}
+	// accessKeySecret is the credential for VMware Workstation Pro REST service, in the form of {username}:{password}
+	return client, nil
+}
+
+func (client MachineVMwareClient) GetMachines() ([]*Machine, error) {
+	machines := []*Machine{}
+	url := "http://localhost:8697/api/vms"
```
Comment (hsluoyz) [inline service/vmware.go]: `http://localhost:8697` should be a configurable field in frontend. Skip this if you don't know how to do it

**Review by hsluoyz — CHANGES_REQUESTED** [review]
Demo video?

File: service/vmware.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+)
+
+type MachineVMwareClient struct {
```
Comment (hsluoyz) [inline service/vmware.go]: Use `Vmware` for all variable naming

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by hsluoyz** [comment]
@Gexinyu2333 rebase your code to resolve the conflicts:

image

**Comment by Gexinyu2333** [comment]
This is a demo video.

https://github.com/user-attachments/assets/158576ba-9429-4a36-8124-8987c68d4dda

**Pull request merged**